D3D12 cleanup

teoxoy opened the pull request:
**Connections**
Follow-up to https://github.com/gfx-rs/wgpu/pull/5956.

**Description**
Mainly cleans-up usage of the D3D12 API.
Also fixes an issue where only high performance adapters would be enumerated if DXGI 1.6 is available.
I will have another follow-up to this that will go over and handle errors on a case-by-case basis (similar to https://github.com/gfx-rs/wgpu/pull/6119).

**Testing**
Existing tests. I also tested DXC locally with the changes to shader compilation.

PR doesn't need to be squashed, each commit builds by itself.

teoxoy committed:
[d3d12] refactor buffer/texture creation functions to return resources instead of using out params

teoxoy committed:
[d3d12] treat failure to populate references to resources as `DeviceError::Unexpected`

These are guaranteed to be populated if the creation methods didn't error. If they are not, the driver/runtime is at fault.

teoxoy committed:
move `hal_usage_error` from the vk backend to lib.rs

teoxoy committed:
[d3d12] make `DxgiLib` and `D3D12Lib` methods consistent

teoxoy committed:
[d3d12] simplify `create_factory()` fn

teoxoy committed:
[d3d12] simplify `enumerate_adapters()` fn

Also fixes an issue where only high performance adapters would be enumerated if DXGI 1.6 is available.

teoxoy committed:
[d3d12] use windows-rs's bindings for DXC instead of hassle-rs

teoxoy committed:
[wgpu-hal] add null terminations to symbols to avoid reallocation

nical reviewed (approved):
Glorious.

teoxoy commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -65,39 +65,72 @@ use crate::auxil::{
         )
-        .map(|| result__.expect("D3D12CreateDevice succeeded but result is NULL?")))
+        .ok()
+        .into_device_result("Device creation")?;
```
This change is causing CI to panic (due to `device_lost_panic`). It seems that creating a device from the WARP adapter fails with `DXGI_ERROR_DEVICE_RESET` in the `DEVICE_LIFETIME_CHECK` & `MULTIPLE_DEVICES` tests.

teoxoy replied:
These 2 tests are not set up properly actually. They create a new instance, adapter and device from env or with defaults. I suspect that when we see they pass on WARP in CI, they actually use llvmpipe (confirmed in https://github.com/gfx-rs/wgpu/pull/6203).

ErichDonGubler replied:
Does this mean we were never properly using WARP adapters? 👀

teoxoy replied:
I'm not sure, creating the first device from the WARP adapter seems to work but creating a 2nd one seems to fail with `DXGI_ERROR_DEVICE_RESET`.

teoxoy replied:
The only thing I could find online that is related to this is:

> Cheaper adapter enumeration
> For Direct3D 12, it’s no longer possible to backtrack from a device to the [IDXGIAdapter](https://learn.microsoft.com/en-us/windows/win32/api/DXGI/nn-dxgi-idxgiadapter) that was used to create it. It’s also no longer possible to provide D3D_DRIVER_TYPE_WARP into [D3D12CreateDevice](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-d3d12createdevice). To make development easier, you can use [IDXGIFactory4](https://learn.microsoft.com/en-us/windows/win32/api/DXGI1_4/nn-dxgi1_4-idxgifactory4) to deal with both of these. [IDXGIFactory4::EnumAdapterByLuid](https://learn.microsoft.com/en-us/windows/win32/api/DXGI1_4/nf-dxgi1_4-idxgifactory4-enumadapterbyluid) (designed to be paired with [ID3D12Device::GetAdapterLuid](https://learn.microsoft.com/en-us/windows/win32/api/d3d12/nf-d3d12-id3d12device-getadapterluid)) enables an app to retrieve information about the adapter where a Direct3D 12 device was created. [IDXGIFactory4::EnumWarpAdapter](https://learn.microsoft.com/en-us/windows/win32/api/DXGI1_4/nf-dxgi1_4-idxgifactory4-enumwarpadapter) provides an adapter which can be provided to D3D12CreateDevice to use the WARP renderer.

_from https://learn.microsoft.com/en-us/windows/win32/direct3ddxgi/dxgi-1-4-improvements#cheaper-adapter-enumeration_

We should run some experiments to see what works and what doesn't. I will open an issue to track this.

For this PR, I modified the tests to reuse the instance/adapter and they are now passing.

teoxoy committed:
[tests] don't create a new instance/adapter for `DEVICE_LIFETIME_CHECK` & `MULTIPLE_DEVICES` tests

The test harness creates the instance and adapter that a test should use. Tests should not create these with default configurations or configs from the environment.

MarijnS95 commented on Cargo.toml:
```diff
@@ -149,7 +149,6 @@ gpu-descriptor = "0.3"
 bit-set = "0.8"
 gpu-allocator = { version = "0.27", default-features = false }
 range-alloc = "0.1"
-hassle-rs = "0.11.0"
```
For future reference, the only reason why `hassle-rs` cannot use `windows-rs` as bindings is because the `windows-rs` maintainers refuse to allow basic Linux support. DirectxShaderConverter works on Linux via the same COM bindings otherwise.

(Either way the abstraction provided by `hassle-rs` is far from ideal)

teoxoy replied:
I found it more straightforward to just use the bindings directly since we have to deal with them for D3D12 anyway.

MarijnS95 replied:
Yeah the abstraction by `hassle-rs` isn't that useful. It's mostly creating the bindings but we have those now (except not for Linux 😭) in `windows-rs`.

MarijnS95 commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -137,8 +167,6 @@ impl D3D12Lib {
         )
         .ok()
-        // TODO: If there's a HRESULT, error may still be non-null and
-        // contain info.
         .into_device_result("Root signature serialization")?;
```
The idea was to check the `error` blob before bailing out with `into_device_result()`, missed this during the previous cleanup. The TODO remains valid.

teoxoy replied:
Hmm, is that something we should handle? What would be an example?

The only scenario I can think of is: the blob getting partially filled with errors after which we get an OOM but it's probably ok to error with an OOM instead of reporting the partial errors.

I wish the exact error codes were documented (like in Vulkan), now we have to guess which variants we might run into...

MarijnS95 replied:
I was expecting a similar setup like `Compile()` the error doesn't need to be OOM.

For `Compile()`, I'm not sure which `HRESULT`s to check either. There is:

- The result from `Compile()`;
- Non-null `error` blob;
- Null `blob`;
- Result from `IDxcOperationResult::GetStatus()`.

(and any getters on `IDxcOperationResult`/`IDxcResult` can fail as well)

teoxoy replied:
Left a reply in https://github.com/gfx-rs/wgpu/pull/6200#discussion_r1744221644.

MarijnS95 commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -147,104 +175,102 @@ impl D3D12Lib {
+        let mut result__ = None;
+
+        (func)(&Direct3D12::ID3D12Debug::IID, <*mut _>::cast(&mut result__))
+            .ok()
+            .into_device_result("GetDebugInterface")?;
 
-        let mut result__ = core::ptr::null_mut();
-        Ok((func)(&Direct3D12::ID3D12Debug::IID, &mut result__)
-            .and_then(|| unsafe { windows_core::Type::from_abi(result__) }))
+        result__.ok_or(crate::DeviceError::Unexpected)
```
Note that a lot of this was intentionally verbatim-copied from `windows-rs` to allow easier updating.

teoxoy replied:
I see, but I think it's more valuable to have those be consistent. I'm also not a fan of using the `from_abi` function which maps null to `Error::empty()`.

MarijnS95 replied:
Right yes, that is unfortunate.

Perhaps this also has to do with SAL annotations, if I remember correctly there are different annotations that allow return values to be `NULL`, versus them only being `NULL` when a non-`SUCCESS` `RESULT` is returned.

The original code now no longer relies on deref semantics, unfortunately.

MarijnS95 commented on wgpu-hal/src/dx12/mod.rs:
```diff
@@ -147,104 +175,102 @@ impl D3D12Lib {
+            &Dxgi::IDXGIFactory4::IID,
+            <*mut _>::cast(&mut result__),
         )
+        .ok()
```
I should report over at `windows-rs` that the naming choice of this function is less than ideal. I keep mistaking it for https://doc.rust-lang.org/stable/std/result/enum.Result.html#method.ok which _discards_ `Error` information by turning `Result` into `Option` which instantly triggers during review.

MarijnS95 commented on wgpu-hal/src/dx12/shader_compilation.rs:
```diff
@@ -1,11 +1,11 @@
+use std::path::PathBuf;
+use windows::{
+    core::{Interface, PCSTR, PCWSTR},
+    Win32::Graphics::Direct3D::{Dxc, Fxc},
+};
 
-pub(super) use dxc::{compile_dxc, get_dxc_container, DxcContainer};
-use windows::Win32::Graphics::Direct3D;
```
Doesn't `wgpu` use `StdExternalCrate` formatting?

teoxoy replied:
It doesn't seem that we do.

MarijnS95 commented on wgpu-hal/src/dx12/shader_compilation.rs:
```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
-                );
-                return Ok(None);
+impl DxcLib {
+    fn new(lib_path: Option<PathBuf>, lib_name: &'static str) -> Result<Self, libloading::Error> {
+        let lib_path = if let Some(lib_path) = lib_path {
+            if lib_path.is_file() {
+                lib_path
+            } else {
+                lib_path.join(lib_name)
             }
+        } else {
+            PathBuf::from(lib_name)
         };
```
I wouldn't have copied this weird bit from `hassle-rs` though, it has caused us quite a few headaches over the years.

teoxoy replied:
I can imagine, but I didn't want to change the behavior (yet) as it is documented in our docs.

MarijnS95 commented on wgpu-hal/src/dx12/shader_compilation.rs:
```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
-                hassle_rs::HassleError::CompileError(_e) => unimplemented!(),
-            }
-        }
+impl OPCWSTR {
+    fn new(s: &str) -> Self {
+        let mut inner: Vec<_> = s.encode_utf16().collect();
+        inner.push(0);
+        Self { inner }
     }
```
Didn't `HSTRING` work for you?

teoxoy replied:
It seemed to be even more complex for this purpose (ref-counted, no easy way to create it from a `&str`, no easy way to get a `PCWSTR` from it).

MarijnS95 commented on wgpu-hal/src/dx12/shader_compilation.rs:
```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
+    kind: Dxc::DXC_OUT_KIND,
+) -> Result<T, crate::DeviceError> {
+    let mut result__: Option<T> = None;
+    unsafe { res.GetOutput::<T>(kind, &mut None, <*mut _>::cast(&mut result__)) }
```
This could coerce because it takes a `*mut Option<T>`

```suggestion
    unsafe { res.GetOutput::<T>(kind, &mut None, &mut result__) }
```

MarijnS95 commented on wgpu-hal/src/dx12/shader_compilation.rs:
```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
+    res: &Dxc::IDxcResult,
+    kind: Dxc::DXC_OUT_KIND,
+) -> Result<T, crate::DeviceError> {
+    let mut result__: Option<T> = None;
```
May just be me who prefers this short-hand:

```suggestion
    let mut result__ = None::<T>;
```

Although the type can be inferred:

```suggestion
    let mut result__ = None;
```

Not that it matters too much, this signature will change when upstream metadata fixes land.

MarijnS95 commented on wgpu-hal/src/dx12/shader_compilation.rs:
```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
+    }
+
+    pub fn create_instance<T: DxcObj>(&self) -> Result<T, crate::DeviceError> {
+        type Fun = extern "system" fn(
+            rclsid: *const windows_core::GUID,
+            riid: *const windows_core::GUID,
+            ppv: *mut *mut core::ffi::c_void,
+        ) -> windows_core::HRESULT;
```
Any reason to not use the existing `DxcCreateInstanceProc` typedef? Because it's wrapped in `Option<>`?

teoxoy replied:
I saw that it exists but wasn't 100% sure how that would behave with libloading. It's not something I've seen in their docs/examples.

MarijnS95 commented on wgpu-hal/src/dx12/shader_compilation.rs:
```diff
@@ -75,234 +73,237 @@ pub(super) fn compile_fxc(
+            .compiler
+            .Compile(&buffer, Some(&compile_args), None)
+    }
+    .into_device_result("Compile")?;
```
Like the old API, I wonder if `Compile` returns an error code while `IDxcResult` is _also_ returned with valid errors (the root cause) in `DXC_OUT_ERRORS`.

Or if that's still communicated by `IDxcOperationResult::GetResult()` which you're not checking here?

teoxoy replied:
From my testing, `IDxcCompiler3::Compile` seems to succeed even if there was a compilation error.

Getting the blob via `IDxcResult::GetOutput(DXC_OUT_ERRORS)` always succeeds. If there is a compilation error its size/length will be non 0 and will contain the error message.

The wiki seems to say that `Compile` will only fail if the compiler itself ran into an internal error.
https://github.com/microsoft/DirectXShaderCompiler/wiki/Advanced-Error-Handling#internal-compiler-errors
